packet/src/bwe/rate_controller.rs

```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    pub(super) fn estimated_bitrate(&self) -> u64 {
+        self.estimated_bitrate.floor() as u64
+    }
+
+    fn increase(&mut self, now: Instant) {
+        let since_last_update = ((now - self.last_estimate_update).as_millis() as f64) / 1000.0;
+
+        let new_estimate = if self.is_near_convergence() {
+            // Additive increase
+            let response_time_ms = self.last_rtt_us.map(|rtt| rtt / 1000.0).unwrap_or(0.0) + 100.0;
+            let alpha = 0.5 * (since_last_update / response_time_ms).min(1.0);
+            let expected_packet_size = self.estimated_packet_size();
+            self.estimated_bitrate + (alpha * expected_packet_size).max(1000.0)
+        } else {
+            // Multiplicative increase
+            let eta = 1.08_f64.powf(since_last_update.min(1.0));
+            eta * self.estimated_bitrate
+        };
+        let max = self.last_observed_bitrate.map(|r| r * 1.5_f64);
+
+        let restrited_estimate = max.map(|m| m.min(new_estimate)).unwrap_or(new_estimate);
+        self.update_estimate(restrited_estimate, now);
+    }
+
+    fn decrease(&mut self, now: Instant) {
+        let new_estimate = BETA * self.estimated_bitrate;
+        if let Some(last_observed_bitrate) = self.last_observed_bitrate {
+            self.averaged_observed_bitrate.update(last_observed_bitrate);
+        }
+        self.update_estimate(new_estimate, now);
+    }
+
+    fn is_near_convergence(&self) -> bool {
+        // Not near convergence until we have valid statistics
+        if !self.averaged_observed_bitrate.valid() {
+            return false;
+        }
+        let Some(last_observed_bitrate) = self.last_observed_bitrate else {
+            return false;
+        };
+
+        // Near convergence if the observed bandwidth is within 3 standard deviations of
+        // the moving average when we have been in the decrease state.
+        self.averaged_observed_bitrate
+            .within_std(last_observed_bitrate, 3.0)
+    }
+
+    fn update_estimate(&mut self, new_estimate: f64, now: Instant) {
+        self.estimated_bitrate = new_estimate.clamp(self.min_bitrate, self.max_bitrate);
+        self.last_estimate_update = now;
+    }
+
+    fn estimated_packet_size(&self) -> f64 {
+        // Assume 30 FPS video dominates the send rate
+        let bits_per_frame = self.estimated_bitrate / 30.0;
+        let packets_per_frame = (bits_per_frame / (1200.0 / 8.0)).ceil();
+
+        bits_per_frame / packets_per_frame
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Signal {
+    Overuse,
+    Underuse,
+    Normal,
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+enum State {
+    Hold,
+    #[default]
+    Increase,
+    Decrease,
+}
+
+impl State {
+    fn transition(&self, signal: Signal) -> Self {
+        match signal {
+            Signal::Overuse => self.overuse(),
+            Signal::Underuse => self.underuse(),
+            Signal::Normal => self.normal(),
```

Comment:
I think this would be more clear as:

```
match (self, signal) {
  (_, Signal::Overuse) => Self::Decrease,
  (_, Signal::Underuse) => Self::Hold,
  (Self::Hold | Self::Increase, Signal::Normal) => Self::Increase,
  (Self::Decrease, Signal::Normal) => Self::Hold,
}
```

Reply:
You're right, I wrote those methods first and then just hammered out this transition method